riscv/src/code_gen.rs

```diff
@@ -319,43 +376,179 @@ fn preamble<T: FieldElement>(runtime: &Runtime, with_bootloader: bool) -> String
+    col witness XX, XX_inv, XXIsZero;
+    std::utils::force_bool(XXIsZero);
+    XXIsZero * XX = 0;
 
     // ============== control-flow instructions ==============
 
-    instr load_label l: label -> X { X = l }
+    // Load the value of label `l` into register X.
+    instr load_label X, l: label
+        link ~> regs.mstore(X, STEP, val1_col)
+    {
+        val1_col = l
+    }
+
+    // Jump to `l` and store the return program counter in register W.
+    instr jump l: label, W
+        link ~> regs.mstore(W, STEP, val3_col)
+    {
+        pc' = l,
+        val3_col = pc + 1
+    }
+    
+    // Jump to the address in register X and store the return program counter in register W.
+    instr jump_dyn X, W
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> regs.mstore(W, STEP, val3_col)
+    {
+        pc' = val1_col,
+        val3_col = pc + 1
+    }
 
-    instr jump l: label -> Y { pc' = l, Y = pc + 1}
-    instr jump_dyn X -> Y { pc' = X, Y = pc + 1}
+    // Jump to `l` if val(X) - val(Y) is nonzero, where X and Y are register ids.
+    instr branch_if_nonzero X, Y, l: label
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> val2_col = regs.mload(Y, STEP + 1)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+        XX = val1_col - val2_col,
+        pc' = (1 - XXIsZero) * l + XXIsZero * (pc + 1)
+    }
 
-    instr branch_if_nonzero X, l: label { pc' = (1 - XIsZero) * l + XIsZero * (pc + 1) }
-    instr branch_if_zero X, l: label { pc' = XIsZero * l + (1 - XIsZero) * (pc + 1) }
+    // Jump to `l` if val(X) - (val(Y) + Z) is zero, where X and Y are register ids and Z is a
+    // constant offset.
+    instr branch_if_zero X, Y, Z, l: label
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> val2_col = regs.mload(Y, STEP + 1)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+        XX = val1_col - (val2_col + Z),
+        pc' = XXIsZero * l + (1 - XXIsZero) * (pc + 1)
+    }
 
-    // Skips Y instructions if X is zero
-    instr skip_if_zero X, Y { pc' = pc + 1 + (XIsZero * Y) }
+    // Skips W instructions if val(X) - val(Y) + Z is zero, where X and Y are register ids and Z is a
+    // constant offset.
+    instr skip_if_zero X, Y, Z, W
+        link ~> val1_col = regs.mload(X, STEP)
+        link ~> val2_col = regs.mload(Y, STEP + 1)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+        XX = val1_col - val2_col + Z,
+        pc' = pc + 1 + (XXIsZero * W)
+    }
 
-    // input X is required to be the difference of two 32-bit unsigend values.
-    // i.e. -2**32 < X < 2**32
-    instr branch_if_positive X, l: label {
-        X + 2**32 - 1 = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000 + wrap_bit * 2**32,
+    // Branches to `l` if V = val(X) - val(Y) + Z is positive, where X and Y are register ids and Z is a
+    // constant offset.
+    // V is required to be the difference of two 32-bit unsigned values.
+    // i.e. -2**32 < V < 2**32.
+    instr branch_if_positive X, Y, Z, l: label
```

Comment:
If it changes the semantics of the instructions I'd rather do it in a different PR because the current semantics were already built/reviewed by 3 people.